docs/design/federated-services.md

```diff
@@ -0,0 +1,490 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+## Cross-cluster Load Balancing and Service Discovery 
+### Requirements and System Design
+### by Quinton Hoole, Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in
+   a Kubernetes cluster) must be able to easily discover and connect
+   to endpoints for Kubernetes services on which they depend in a
+   consistent way, irrespective of whether those services exist in a
+   different kubernetes cluster within the same cluster federation.
+   Hence-forth referred to as "cluster-internal clients", or simply
+   "internal clients".
+1. **External discovery and connection**: External clients (running
+   outside a Kubernetes cluster) must be able to discover and connect
+   to endpoints for Kubernetes services on which they depend.
+   1. **External clients predominantly speak HTTP(S)**: External
+      clients are most often, but not always, web browsers, or at
+      least speak HTTP(S) (TBD list notable exceptions.  DNS servers?
+      SIP servers?  Database Servers?)
+1. **Find the "best" endpoint:** Upon initial discovery and
```

Comment:
Hmm, I think you're placing un-necessary emphasis on the geographic routing aspect. If you simply choose an endpoint based on health check rtt, won't that suffice (or rather, you simply expose whatever loadbalancing algorightms the underlying lb supports and let the user pick one)? 

There are cases where one might really want the *closest* endpoint (eg: you want to terminate ssl closest to the user because of TLS negotiation and then send http bouncing wherever there is capacity), and if that's what you're prioritizing, please be explicit.